### PR TITLE
issue a warning when Item and Collector are used in diamond inheritance

### LINKED ISSUE #8435: Unexpected keyword argument 'path' from plugins
While troubleshooting #8332, I stumbled onto a new error, a `TypeError` that occurs when using pytest-black against the current main HEAD (32ad70d), easily reproducible with an empty test file and pip-run:

```
draft $ touch test_something.py
draft $ pip-run -q git+https://github.com/pytest-dev/pytest pytest-black -- -m pytest --black
===================================================================================== test session starts =====================================================================================
platform darwin -- Python 3.9.2, pytest-6.3.0.dev252+g32ad70dea, py-1.10.0, pluggy-0.13.1
rootdir: /Users/jaraco/draft
plugins: black-0.3.12
collected 0 items / 1 error                                                                                                                                                                   

=========================================================================================== ERRORS ============================================================================================
________________________________________________________________________________ ERROR collecting test session ________________________________________________________________________________
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-j9xn8e36/pluggy/hooks.py:286: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-j9xn8e36/pluggy/manager.py:93: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-j9xn8e36/pluggy/manager.py:84: in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-j9xn8e36/pytest_black.py:27: in pytest_collect_file
    return BlackItem.from_parent(parent, fspath=path)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-j9xn8e36/_pytest/nodes.py:578: in from_parent
    return super().from_parent(parent=parent, fspath=fspath, path=path, **kw)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-j9xn8e36/_pytest/nodes.py:226: in from_parent
    return cls._create(parent=parent, **kw)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-j9xn8e36/_pytest/nodes.py:117: in _create
    return super().__call__(*k, **kw)
E   TypeError: __init__() got an unexpected keyword argument 'path'
=================================================================================== short test summary info ===================================================================================
ERROR  - TypeError: __init__() got an unexpected keyword argument 'path'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
====================================================================================== 1 error in 0.13s =======================================================================================
```

Same problem happens with pytest-checkdocs:

```
draft $ touch setup.py
draft $ pip-run -q git+https://github.com/pytest-dev/pytest pytest-checkdocs -- -m pytest
===================================================================================== test session starts =====================================================================================
platform darwin -- Python 3.9.2, pytest-6.3.0.dev252+g32ad70dea, py-1.10.0, pluggy-0.13.1
rootdir: /Users/jaraco/draft
plugins: checkdocs-2.4.0
collected 0 items / 1 error                                                                                                                                                                   

=========================================================================================== ERRORS ============================================================================================
________________________________________________________________________________ ERROR collecting test session ________________________________________________________________________________
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/pluggy/hooks.py:286: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/pluggy/manager.py:93: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/pluggy/manager.py:84: in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/pytest_checkdocs/__init__.py:14: in pytest_collect_file
    CheckdocsItem.from_parent(parent, fspath=path)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/pytest_checkdocs/__init__.py:52: in from_parent
    return super().from_parent(parent, fspath=fspath)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/_pytest/nodes.py:578: in from_parent
    return super().from_parent(parent=parent, fspath=fspath, path=path, **kw)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/_pytest/nodes.py:226: in from_parent
    return cls._create(parent=parent, **kw)
/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-run-5hc5__bb/_pytest/nodes.py:117: in _create
    return super().__call__(*k, **kw)
E   TypeError: __init__() got an unexpected keyword argument 'path'
=================================================================================== short test summary info ===================================================================================
ERROR  - TypeError: __init__() got an unexpected keyword argument 'path'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
====================================================================================== 1 error in 0.20s =======================================================================================
```

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
resolves #8435

- [x] include doc example from https://github.com/asmeurer/pytest-flakes/pull/40/files
- [x] document the problem in more detail
- [x] consider the addition of a `FileItem` with unique test ids
- [x] return ctor backward compat

### CODE REVIEW

File: testing/test_nodes.py [hunk 1]
```diff
@@ -38,6 +38,16 @@ def test_node_from_parent_disallowed_arguments() -> None:
         nodes.Node.from_parent(None, config=None)  # type: ignore[arg-type]
 
 
+def test_subclassing_node_with_item_warns() -> None:
```
Comment (bluetech) [inline testing/test_nodes.py:41]:
```suggestion
def test_subclassing_both_item_and_collector_warns() -> None:
```

File: changelog/8447.deprecation.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Warn when inheritance is used to bring Collectors into Items,
+it was never sanely supported and triggers hard to debug errors.
+
+Instead of inheritance composition should be used.
```
Comment (bluetech) [inline changelog/8447.deprecation.rst:4]:
```suggestion
Defining a custom pytest node type which is both an item and a collector now issues a warning.
It was never sanely supported and triggers hard to debug errors.

Instead, a separate collector node should be used, which collects the item. See :ref:`nonpython` for an example.
```

File: src/_pytest/nodes.py [hunk 3]
```diff
@@ -601,6 +602,19 @@ class Item(Node):
 
     nextitem = None
 
+    def __init_subclass__(cls):
```
Comment (bluetech) [inline src/_pytest/nodes.py:605]:
```suggestion
    def __init_subclass__(cls) -> None:
```

**Review by bluetech — COMMENTED** [review]
I think it's a good idea to warn about it, so +1 from me.

I left some suggestions on the wording.

**Comment by RonnyPfannschmidt** [comment]
will complete the docs tommorow morning

**Comment by The-Compiler** [comment]
CI failures (other than the `docs` failure) are unrelated, see https://github.com/codecov/codecov-action/issues/330.

FWIW we've had a plugin at work which had the same issue as in #8435, but without any diamond inheritance involved. It looks like this has been resolved somewhere along the way though, as I now don't see any issues with either the `main` branch nor this PR and the plugin.

Other than that, I'll let someone else review, as I'm not too familiar with this part of the code.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Just a couple of things from a quick look.

File: src/_pytest/nodes.py [hunk 4]
```diff
@@ -125,7 +128,14 @@ def __call__(self, *k, **kw):
         fail(msg, pytrace=False)
 
     def _create(self, *k, **kw):
-        return super().__call__(*k, **kw)
+        try:
+            return super().__call__(*k, **kw)
+        except TypeError:
+            sig = signature(cast(Type[Node], self).__init__)
+            sig.replace()
```
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py:135]: good catch,

i initially removed self and used binding (but noted that the binding was just not working out as i liked)

its no longer used

File: src/_pytest/nodes.py [hunk 5]
```diff
@@ -125,7 +128,14 @@ def __call__(self, *k, **kw):
         fail(msg, pytrace=False)
 
     def _create(self, *k, **kw):
-        return super().__call__(*k, **kw)
+        try:
+            return super().__call__(*k, **kw)
+        except TypeError:
+            sig = signature(cast(Type[Node], self).__init__)
```
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py:134]: yeah, that might be a good call in that particular location

File: src/_pytest/nodes.py [hunk 6]
```diff
@@ -125,7 +128,14 @@ def __call__(self, *k, **kw):
...
+            sig.replace()
+            known_kw = {k: v for k, v in kw.items() if k in sig.parameters}
```
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py:136]: modern constructors are expected to pass over unknown arguments
legacy/broken ones did not

but an extra warning should be added

[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py:135]: done

File: src/_pytest/nodes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py]: getattr actually works and is even better

File: src/_pytest/nodes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py]: fixed

File: src/_pytest/nodes.py [hunk 7]
```diff
@@ -610,15 +644,36 @@ class Item(Node):
...
+    def __init_subclass__(cls) -> None:
+        problems = ", ".join(
+            base.__name__ for base in cls.__bases__ if issubclass(base, Collector)
+        )
+        if problems:
+            warnings.warn(
+                f"{cls.__name__} is an Item subclass and should not be a collector, "
+                f"however its bases {problems} are collectors.\n"
+                "Please split the Collectors and the Item into separate node types.\n"
+                "TODO: doc link",
```
Comment (nicoddemus) [inline src/_pytest/nodes.py:656]: TODO left here.

**Review by nicoddemus — APPROVED** [review]
Overall looks good, seems you missed a TODO in the code. 👍

File: src/_pytest/nodes.py (hunk 7 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/nodes.py]: done thanks

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt just a ping in case you missed it, but a test needs update it seems. 👍

**Comment by RonnyPfannschmidt** [comment]
Will fix as soon as I can access a computer

**Comment by bluetech** [comment]
I understand the `__init_subclass__` change (which LGTM), but I don't understand the other changes, to e.g. `FSCollector` and the constructors, and how they are related to the diamond inheritance issue, can you explain?

**Comment by RonnyPfannschmidt** [comment]
They restore the ability to use the pattern, i accidentally broke it before (which caused issues to projects)

**Review by The-Compiler — CHANGES_REQUESTED** [review]
I'll have to agree that I don't really follow about the `FSCollector` changes unfortunately. We should probably at least have a test demonstrating what worked with 6.2, breaks with `main`, but works again with this PR?

**Comment by RonnyPfannschmidt** [comment]
The test added does that

In particular the constructor used, 

i will attach a log once I return

**Comment by RonnyPfannschmidt** [comment]
@The-Compiler 


pytest on main with the test having the warnings removed will do the following

test: ```python


def test_subclassing_both_item_and_collector_deprecated(request, tmp_path: Path) -> None:
    """
    verifies we warn on diamond inheritance
    as well as correctly managing legacy inheritance ctors with missing args
    as found in plugins
    """

    class SoWrong(nodes.File, nodes.Item):
        def __init__(self, fspath, parent):
            """Legacy ctor with legacy call # don't wana see"""
            super().__init__(fspath, parent)

    SoWrong.from_parent(
        request.session, fspath=legacy_path(tmp_path / "broken.txt")
    )
```

out: ```
$ pytest testing/test_nodes.py -k test_subclassing_both_item_and_collector_deprecated --tb=native
================================================================================================================================= test session starts ==================================================================================================================================
platform linux -- Python 3.9.5, pytest-6.3.0.dev515+g46269d90d.d20210617, py-1.10.0, pluggy-0.13.1
rootdir: /home/ronny/Projects/pytest-dev/pytest, configfile: pyproject.toml
collected 17 items / 16 deselected / 1 selected                                                                                                                                                                                                                                        

testing/test_nodes.py F                                                                                                                                                                                                                                                          [100%]

======================================================================================================================================= FAILURES =======================================================================================================================================
_________________________________________________________________________________________________________________ test_subclassing_both_item_and_collector_deprecated __________________________________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/ronny/Projects/pytest-dev/pytest/testing/test_nodes.py", line 56, in test_subclassing_both_item_and_collector_deprecated
    SoWrong.from_parent(
  File "/home/ronny/Projects/pytest-dev/pytest/src/_pytest/nodes.py", line 587, in from_parent
    return super().from_parent(parent=parent, fspath=fspath, path=path, **kw)
  File "/home/ronny/Projects/pytest-dev/pytest/src/_pytest/nodes.py", line 237, in from_parent
    return cls._create(parent=parent, **kw)
  File "/home/ronny/Projects/pytest-dev/pytest/src/_pytest/nodes.py", line 128, in _create
    return super().__call__(*k, **kw)
TypeError: __init__() got an unexpected keyword argument 'path'
=============================================================================================================================== short test summary info ================================================================================================================================
FAILED testing/test_nodes.py::test_subclassing_both_item_and_collector_deprecated - TypeError: __init__() got an unexpected keyword argument 'path'
=========================================================================================================================== 1 failed, 16 deselected in 0.01s ===========================================================================================================================
```
with the extra support code it passes

**Comment by RonnyPfannschmidt** [comment]
@The-Compiler done

**Review by The-Compiler — APPROVED** [review]
I still really can't claim to fully understand the topic, but I played around a bit with [partially reverting](https://paste.the-compiler.org/view/82f296a9) changes and indeed [stuff breaks](https://paste.the-compiler.org/view/8c80ae56) when doing that.

Given that @nicoddemus was fine with it earlier and there aren't any other comments, let's finally get this in! :tada: Thanks @RonnyPfannschmidt for the persistence!

**Comment by RonnyPfannschmidt** [comment]
@The-Compiler this one is a real backward compatibility pain, i hope we can take the mechanisms behind it out of the picture soon

i'm not happy we have to have this workaround for now

**Pull request merged**